### DISCUSSION TITLE
FIPS configuration for Quarkus distribution

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#12250] [Archive]
To enable FIPS compliance in Keycloak there's 3 things that needs to be configured

1. Security providers in JVM
2. TLS for Vert.X
3. BouncyCastle

With regard to 1 if using RHEL or Fedora this can be achieved by enabling FIPS at the host level, which automatically configures OpenJDK. Ideally we'd detect this in Keycloak and automatically configure FIPS in Keycloak when enabled at the host level. For other distributions we'd need some example/documentation on configuring the security providers in OpenJDK, and potentially a config option to enable FIPS mode in Keycloak.

With regard to 2 it may be sufficient to enable FIPS at the OpenJDK level without any additional configuration of Quarkus. This needs to be investigated though.

The bigger topic as part of this discussion is how we achieve 3. BouncyCastle provides separate JARs for BouncyCastle FIPS, including a different security provider. We'd probably want to distribution the Quarkus distribution both with BouncyCastle and BouncyCastle FIPS, and switch which is used with a config option. Ideal would also be to only load the needed JARs on the classpath.

We'll have separate work to update the code-base to allow FIPS compliance, this discussion is limited to configuration at the Quarkus dist level.

**Comment by stianst** [comment]
@mposolda  @pedroigor FIY

**Comment by pedroigor** [comment]
If we ship both artifacts in the distribution it should be possible to have only one of them at runtime. However, this would mean having FIPS as a build option.

W.r.t. to #2 do you mean https://quarkus.io/guides/security-customization? And how the store providers are set?

**Reply by stianst** [reply]
FIPS at build-time works perfectly. One alternative would be to just use BouncyCastle FIPS libraries always, not sure if that's a good idea or not though.

**Reply by stianst** [reply]
For #2 not sure if we need to do anything, as I presume once you have an OpenJDK with FIPS compliant security providers Quarkus at the https level will be FIPS compliant, and there's no need to change to use BouncyCastle FIPS. Not 100% sure on this one, perhaps something we should check with Quarkus folks.

**Reply by pedroigor** [reply]
We can also start a PoC around it. I have a feeling that the only thing we need is to wrap the `quarkus.security.security-provider` to set the choosen provider (e.g.: `BCFIPS`) and expect Quarkus to use the corresponding provider when resolving algorithms. Of course, plus the artifacts either in the dist classpath or within the `providers` dir.

**Reply by pedroigor** [reply]
> FIPS at build-time works perfectly. One alternative would be to just use BouncyCastle FIPS libraries always, not sure if that's a good idea or not though.

I think we can do it and avoid an additional step to copy libraries when doing FIPS. As I mentioned, whe re-aug we should be able to exclude artifacts from the runtime classpath.

**Reply by stianst** [reply]
@pedroigor I think there may be some reasons why we can't actually distribute the BC FIPS with Keycloak, at least for prod side. So need to check if we even can include BC FIPS libs out of the box.

**Reply by stianst** [reply]
> We can also start a PoC around it. I have a feeling that the only thing we need is to wrap the `quarkus.security.security-provider` to set the choosen provider (e.g.: `BCFIPS`) and expect Quarkus to use the corresponding provider when resolving algorithms. Of course, plus the artifacts either in the dist classpath or within the `providers` dir.

I'm thinking if we don't pass anything to Quarkus in this regard it'll just use the security providers in OpenJDK, which will already be FIPS compliant, so no need to make it use BC

**Reply by stianst** [reply]
By the way one more option could maybe be to eliminate BC altogether. Not sure if that's feasible though.

**Comment by stianst** [comment]
Another question is (assuming we are still using BouncyCastle) if there are any API difference needed (ideal is if there isn't and the KC code can work with both BC and BC-FIPS) how do we tackle that. The ideal way could be to create a Crypto SPI and Crypto Provider wrapping the BC code. Then we can have a BC and a BC-FIPS provider implementation. This would also allow the module in the codebase to have the correct dependencies so we can make sure the API is compatible.
